android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFirebaseCoreService.kt

```diff
@@ -0,0 +1,222 @@
+import expo.modules.manifests.core.Manifest
+import host.exp.exponent.kernel.ExperienceKey
+import org.json.JSONObject
+import versioned.host.exp.exponent.modules.universal.ScopedFirebaseCoreService.Companion.addJSONStringToMap
+import java.io.UnsupportedEncodingException
+import java.lang.Exception
+import java.util.*
+
+class ScopedFirebaseCoreService(
+  context: Context,
+  manifest: Manifest,
+  experienceKey: ExperienceKey
+) : FirebaseCoreService(context), RegistryLifecycleListener {
+  private val appName: String
+  private val appOptions: FirebaseOptions?
+
+  override fun getAppName(): String {
+    return appName
+  }
+
+  override fun getAppOptions(): FirebaseOptions? {
+    return appOptions
+  }
+
+  override fun isAppAccessible(name: String): Boolean {
+    synchronized(protectedAppNames) {
+      if (protectedAppNames.containsKey(name) && appName != name) {
+        return false
+      }
+    }
+    return super.isAppAccessible(name)
+  }
+
+  // Registry lifecycle events
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    // noop
+  }
+
+  override fun onDestroy() {
+    // Mark this Firebase App as deleted. Don't delete it straight
+    // away, but mark it for deletion. When loading a new project
+    // a check is performed that will cleanup the deleted Firebase apps.
+    // This ensures that Firebase Apps don't get deleted/recreated
+    // every time a project reload happens, and also also ensures that
+    // `isAppAccessible` keeps the app unavailable for other project/packages
+    // after unload.
+    synchronized(protectedAppNames) { protectedAppNames.put(appName, true) }
+  }
+
+  companion object {
+    private val protectedAppNames = mutableMapOf<String, Boolean>() // Map<App-name, isDeleted>
+
+    private fun getEncodedExperienceScopeKey(experienceKey: ExperienceKey): String {
+      return try {
+        val encodedUrl = experienceKey.getUrlEncodedScopeKey()
+        val data = encodedUrl.toByteArray(charset("UTF-8"))
+        Base64.encodeToString(
+          data,
+          Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP
+        )
+      } catch (e: UnsupportedEncodingException) {
+        experienceKey.scopeKey.hashCode().toString()
+      }
+    }
+
+    // google-services.json loading
+
+    private fun getJSONStringByPath(jsonArg: JSONObject?, path: String): String? {
+      var json = jsonArg ?: return null
+      return try {
+        val paths = path.split(".").toTypedArray()
+        for (i in paths.indices) {
+          val name = paths[i]
+          if (!json.has(name)) return null
+          if (i == paths.size - 1) {
+            return json.getString(name)
+          } else {
+            json = json.getJSONObject(name)
+          }
+        }
+        null
+      } catch (err: Exception) {
+        null
+      }
+    }
+
+    private fun MutableMap<String, String>.addJSONStringToMap(
+      json: JSONObject?,
+      path: String,
+      name: String
+    ) {
```

Comment:
Lines up better with MutableMap's other APIs

```suggestion
    private fun MutableMap<String, String>.putJSONString(
      key: String,
      json: JSONObject?,
      path: String
    ) {
```